db/compaction/compaction_iterator.cc

```diff
@@ -204,72 +204,130 @@ void CompactionIterator::Next() {
 
 bool CompactionIterator::InvokeFilterIfNeeded(bool* need_skip,
                                               Slice* skip_until) {
-  if (compaction_filter_ != nullptr &&
-      (ikey_.type == kTypeValue || ikey_.type == kTypeBlobIndex)) {
-    // If the user has specified a compaction filter and the sequence
-    // number is greater than any external snapshot, then invoke the
-    // filter. If the return value of the compaction filter is true,
-    // replace the entry with a deletion marker.
-    CompactionFilter::Decision filter;
-    compaction_filter_value_.clear();
-    compaction_filter_skip_until_.Clear();
-    CompactionFilter::ValueType value_type =
-        ikey_.type == kTypeValue ? CompactionFilter::ValueType::kValue
-                                 : CompactionFilter::ValueType::kBlobIndex;
-    // Hack: pass internal key to BlobIndexCompactionFilter since it needs
-    // to get sequence number.
-    Slice& filter_key = ikey_.type == kTypeValue ? ikey_.user_key : key_;
-    {
-      StopWatchNano timer(clock_, report_detailed_time_);
+  if (!compaction_filter_ ||
+      (ikey_.type != kTypeValue && ikey_.type != kTypeBlobIndex)) {
+    return true;
+  }
+  bool error = false;
+  // If the user has specified a compaction filter and the sequence
+  // number is greater than any external snapshot, then invoke the
+  // filter. If the return value of the compaction filter is true,
+  // replace the entry with a deletion marker.
+  CompactionFilter::Decision filter = CompactionFilter::Decision::kUndetermined;
+  compaction_filter_value_.clear();
+  compaction_filter_skip_until_.Clear();
+  CompactionFilter::ValueType value_type =
+      ikey_.type == kTypeValue ? CompactionFilter::ValueType::kValue
+                               : CompactionFilter::ValueType::kBlobIndex;
+  // Hack: pass internal key to BlobIndexCompactionFilter since it needs
+  // to get sequence number.
+  assert(compaction_filter_);
+  Slice& filter_key = (ikey_.type == kTypeValue ||
+                       !compaction_filter_->IsStackedBlobDbCompactionFilter())
+                          ? ikey_.user_key
+                          : key_;
+  {
+    StopWatchNano timer(clock_, report_detailed_time_);
+    if (kTypeBlobIndex == ikey_.type) {
+      filter = compaction_filter_->ShouldFilterBlobByKey(
+          compaction_->level(), filter_key, &compaction_filter_value_,
+          compaction_filter_skip_until_.rep());
```

Comment:
Fair point. My thoughts: for regular value, there does not seem a need to separate the filtering logic into two steps, aka filter-by-key and filter-by-key-and-value, since we can call `FilterV2()` for regular value with both key and value available. This is at most one function call. If we call `ShouldFilterByKey()` for regular values, then we may end up with two function calls in the worst case.

Reply:
If users can actually solely rely on keys to perform filtering for all the data, the implementation of their compaction filter can still avoid code duplication. They do not have to implement the same logic twice in `FilterV2()` and `ShouldFilterBlobByKey()`. They can have a common helper function which can be called in both of the above.

Reply:
Yes, good point. Could we still rename the method though? "Should", in particular, throws me off a bit; to me at least that would suggest a yes/no decision about the necessity of filtering, while this method actually performs the filtering in pretty much the same way as `FilterV2`.

I think it would also make sense to mention which KVs it is called for in `compaction_filter.h`.